src/Adapter/Product/Update/ProductShopUpdater.php

```diff
@@ -0,0 +1,306 @@
+        // (e.g copy only prices but not stock)
+        $fields = $fields + $this->getAssociationFields($sourceProduct);
+        $fields = $fields + $this->getPricesFields($sourceProduct);
+        $fields = $fields + $this->getStockFields($sourceProduct);
+        $fields = $fields + $this->getSEOFields($sourceProduct);
+        $fields = $fields + $this->getOptionsFields($sourceProduct);
+        $fields = $fields + $this->getCustomizationFields($sourceProduct);
+        $fields = $fields + $this->getDateFields($sourceProduct);
+
+        if ($this->productRepository->isAssociatedToShop($productId, $targetShopId)) {
+            $qb = $this->getUpdateQueryBuilder($fields, $productId, $targetShopId);
+        } else {
+            $qb = $this->getInsertQueryBuilder($fields, $productId, $targetShopId);
+        }
+
+        foreach ($fields as $column => $value) {
+            $qb->setParameter($column, $value);
+        }
+        $qb->execute();
+    }
+
+    /**
+     * @param array $fields
+     * @param ProductId $productId
+     * @param ShopId $targetShopId
+     *
+     * @return QueryBuilder
+     */
+    private function getInsertQueryBuilder(
+        array $fields,
+        ProductId $productId,
+        ShopId $targetShopId
+    ): QueryBuilder {
+        $fields['id_product'] = $productId->getValue();
+        $fields['id_shop'] = $targetShopId->getValue();
+
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->insert($this->dbPrefix . 'product_shop')
+            ->values($this->formatQueryValues($fields))
+            ->setParameter('id_product', $productId->getValue())
+            ->setParameter('id_shop', $targetShopId->getValue())
+        ;
+
+        return $qb;
+    }
+
+    /**
+     * @param array $fields
+     * @param ProductId $productId
+     * @param ShopId $targetShopId
+     *
+     * @return QueryBuilder
+     */
+    private function getUpdateQueryBuilder(
+        array $fields,
+        ProductId $productId,
+        ShopId $targetShopId
+    ): QueryBuilder {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->update($this->dbPrefix . 'product_shop')
+            ->where('id_shop = :id_shop')
+            ->where('id_product = :id_product')
+            ->setParameter('id_shop', $targetShopId->getValue())
+            ->setParameter('id_product', $productId->getValue())
+        ;
+
+        $queryValues = $this->formatQueryValues($fields);
+        foreach ($queryValues as $column => $value) {
+            $qb->set($column, $value);
+        }
+
+        return $qb;
+    }
+
+    /**
+     * @param Product $product
+     *
+     * @return array
+     */
+    private function getAssociationFields(Product $product): array
+    {
+        $associationFields = [];
+        $associationFields['id_category_default'] = (int) $product->id_category_default;
+        $associationFields['cache_default_attribute'] = (int) $product->cache_default_attribute;
```

Comment:
If you are going this way, you can do it with less noise:
```suggestion
        $associationFields = [
        	'id_category_default' => (int) $product->id_category_default,
        	'cache_default_attribute' => (int) $product->cache_default_attribute,
      	];
```